File: src/main/java/swordofmagic7/Attribute.java
package swordofmagic7;

import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;

import static swordofmagic7.DataBase.getClassList;
import static swordofmagic7.DataBase.playerData;
import static swordofmagic7.Function.*;

enum AttributeType {
    STR("§c§l筋力", Material.RED_DYE, "§a§l物理攻撃に関するステータスに影響します"),
    INT("§d§l魔力", Material.PURPLE_DYE, "§a§l魔法攻撃に関するステータスに影響します"),
    DEX("§e§l敏捷", Material.YELLOW_DYE, "§a§l回避とクリティカルダメージに影響します"),
    TEC("§2§l技量", Material.GREEN_DYE, "§a§lクリティカル発生と命中に関するステータスに影響します"),
    SPI("§b§l精神", Material.LIGHT_BLUE_DYE, "§a§lマナと魔法防御に関するステータスに影響します"),
    VIT("§6§l活力", Material.ORANGE_DYE, "§a§l体力と防御に関するステータスに影響します"),
    ;

    String Display;
    Material Icon;
    List<String> Lore;

    AttributeType(String Display, Material Icon, String Lore) {
        this.Display = Display;
        this.Icon = Icon;
        this.Lore = List.of(Lore.split("\n"));
    }
}

public class Attribute {
    private final Player player;
    private final PlayerData playerData;
    private final HashMap<AttributeType, Integer> Parameter = new HashMap<>();
    private int AttributePoint = 0;

    Attribute(Player player, PlayerData playerData) {
        this.player = player;
        this.playerData = playerData;
        for (AttributeType attr : AttributeType.values()) {
            Parameter.put(attr, 0);
        }
    }

    HashMap<AttributeType, Integer> getAttribute() {
        return Parameter;
    }

    int getAttributePoint() {
        return AttributePoint;
    }

    int getAttribute(AttributeType type) {
        return Parameter.get(type);
    }

    void addPoint(int add) {
        AttributePoint += add;
    }
    void setPoint(int point) {
        AttributePoint = point;
    }

    void addAttribute(AttributeType type, int add) {
        if (getAttributePoint() >= add) {
            AttributePoint -= add;
            Parameter.put(type, Parameter.get(type) + add);
        } else {
            player.sendMessage("§eポイント§aが足りません");
        }
    }

    void setAttribute(AttributeType type, int attr) {
        Parameter.put(type, attr);
    }

    void resetAttribute() {
        for (AttributeType attr : AttributeType.values()) {
            Parameter.put(attr, 0);
        }
        int point = 0;
        for (ClassData classData : getClassList().values()) {
            point += playerData.Classes.getLevel(classData)-1;
        }
        AttributePoint = point;
    }

    ItemStack attributeView(AttributeType type) {
        ItemStack item = new ItemStack(type.Icon);
        ItemMeta meta = item.getItemMeta();
        meta.setDisplayName(decoText(type.Display + " §e§l[" + Parameter.get(type) + "] "));
        List<String> Lore = new ArrayList<>(type.Lore);
        Lore.add(decoText("§3§l追加ステータス"));
        final String format = "%.1f";
        if (type == AttributeType.STR) {
            Lore.add(decoLore("物理与ダメージ") + "+" + String.format(format, Parameter.get(type)*0.5) + "%");
            Lore.add(decoLore("攻撃力") + "+" + String.format(format, Parameter.get(type)*0.5) + "%");
        } else if (type == AttributeType.INT) {
            Lore.add(decoLore("魔法与ダメージ") + "+" + String.format(format, Parameter.get(type)*0.4) + "%");
            Lore.add(decoLore("魔法被ダメージ軽減") + "+" + String.format(format, Parameter.get(type) * 0.1) + "%");
            Lore.add(decoLore("攻撃力") + "+" + String.format(format, Parameter.get(type)*0.5) + "%");
        } else if (type == AttributeType.DEX) {
            Lore.add(decoLore("回避") + "+" + String.format(format, Parameter.get(type)*0.8) + "%");
            Lore.add(decoLore("クリティカルダメージ") + "+" + String.format(format, Parameter.get(type)*0.8) + "%");
        } else if (type == AttributeType.TEC) {
            Lore.add(decoLore("命中") + "+" + String.format(format, Parameter.get(type)*0.8) + "%");
            Lore.add(decoLore("クリティカル発生") + "+" + String.format(format, Parameter.get(type)*0.8) + "%");
        } else if (type == AttributeType.SPI) {
            Lore.add(decoLore("最大マナ") + "+" + String.format(format, Parameter.get(type) * 0.8) + "%");
            Lore.add(decoLore("マナ自動回復") + "+" + String.format(format, Parameter.get(type) * 0.6) + "%");
            Lore.add(decoLore("防御力") + "+" + String.format(format, Parameter.get(type) * 0.2) + "%");
            Lore.add(decoLore("クリティカル耐性") + "+" + String.format(format, Parameter.get(type)*0.2) + "%");
            Lore.add(decoLore("魔法被ダメージ軽減") + "+" + String.format(format, Parameter.get(type) * 0.1) + "%");
        } else if (type == AttributeType.VIT) {
            Lore.add(decoLore("最大体力") + "+" + String.format(format, Parameter.get(type) * 0.8) + "%");
            Lore.add(decoLore("体力自動回復") + "+" + String.format(format, Parameter.get(type) * 0.2) + "%");
            Lore.add(decoLore("防御力") + "+" + String.format(format, Parameter.get(type) * 0.5) + "%");
            Lore.add(decoLore("物理被ダメージ軽減") + "+" + String.format(format, Parameter.get(type) * 0.3) + "%");
            Lore.add(decoLore("魔法被ダメージ軽減") + "+" + String.format(format, Parameter.get(type) * 0.1) + "%");
        }
        meta.setLore(Lore);
        item.setItemMeta(meta);
        return item;
    }
}

File: src/main/kotlin/net/somrpg/swordofmagic7/player/attribute/Attribute.kt
package net.somrpg.swordofmagic7.player.attribute

import io.papermc.paper.datacomponent.DataComponentTypes
import io.papermc.paper.datacomponent.item.ItemLore
import net.kyori.adventure.text.Component
import org.bukkit.Material
import org.bukkit.entity.Player
import org.bukkit.event.inventory.ClickType
import org.bukkit.inventory.Inventory
import org.bukkit.inventory.InventoryView
import org.bukkit.inventory.ItemStack
import swordofmagic7.Data.DataBase
import swordofmagic7.Data.PlayerData
import swordofmagic7.Data.TitleData
import swordofmagic7.Function
import swordofmagic7.Item.ItemStackData
import swordofmagic7.Menu.Data
import swordofmagic7.Sound.CustomSound
import swordofmagic7.Sound.SoundList
import java.text.NumberFormat
import java.util.EnumMap

class Attribute(
    private val player: Player,
    private val playerData: PlayerData,
) {
    companion object {
        private val numberFormat =
            NumberFormat.getNumberInstance().apply {
                maximumFractionDigits = 2
            }
    }

    val attributeMap = EnumMap<AttributeType, Int>(AttributeType::class.java)
    var attributePoint: Int = 0
        private set

    fun getAttribute(type: AttributeType?): Int = attributeMap[type]!!

    fun addPoint(add: Int) {
        this.attributePoint += add
    }

    fun setPoint(point: Int) {
        this.attributePoint = point
    }

    fun addAttribute(
        type: AttributeType?,
        add: Int,
    ) {
        if (this.attributePoint >= add) {
            this.attributePoint -= add
            attributeMap[type] = attributeMap[type]!! + add
        } else {
            player.sendMessage("§eポイント§aが足りません")
        }
    }

    fun revAttribute(
        type: AttributeType?,
        rev: Int,
    ) {
        if (attributeMap[type]!! >= rev) {
            this.attributePoint += rev
            attributeMap[type] = attributeMap[type]!! - rev
        } else {
            player.sendMessage("§eポイント§aが足りません")
        }
    }

    fun setAttribute(
        type: AttributeType?,
        attr: Int,
    ) {
        attributeMap[type] = attr
    }

    fun resetAttribute() {
        for (attr in AttributeType.entries) {
            attributeMap[attr] = 0
        }
        this.attributePoint = (playerData.Level - 1) * 5
        for (data in playerData.titleManager.TitleList) {
            val titleData: TitleData = DataBase.TitleDataList[data]!!
            this.attributePoint += titleData.attributePoint
        }
    }

    @Suppress("UnstableApiUsage")
    fun attributeView(type: AttributeType): ItemStack {
        val item = ItemStack.of(type.icon)
        val lore = mutableListOf<String>()
        lore.add(type.lore)
        lore.add(Function.decoText("§3§l追加ステータス"))
        for (stat in type.stats) {
            val value = stat.value * attributeMap[type]!!
            lore.add("${Function.decoLore(stat.key.Display)}+${numberFormat.format(value)}%")
        }
        item.setData(
            DataComponentTypes.ITEM_NAME,
            Component.text(Function.decoText(type.display + " §e§l[" + attributeMap[type] + "] ")),
        )
        item.setData(DataComponentTypes.LORE, ItemLore.lore(lore.map { Component.text(it) }))
        return item
    }

    private var attributeMenuCache: Inventory? = null

    init {
        for (attr in AttributeType.entries) {
            attributeMap[attr] = 0
        }
    }

    fun attributeMenuView() {
        attributeMenuCache = Function.decoInv(Data.AttributeMenuDisplay, 1)
        attributeMenuLoad()
        player.openInventory(attributeMenuCache!!)
    }

    fun attributeMenuLoad() {
        val attribute = playerData.Attribute
        for ((slot, attr) in AttributeType.entries.withIndex()) {
            attributeMenuCache!!.setItem(slot, attribute.attributeView(attr))
        }
        val lore: MutableList<String?> = ArrayList()
        lore.add(Function.decoLore("ポイント") + attribute.attributePoint)
        lore.add("")
        lore.add("§c§l※クリックでアトリビュートをリセット")
        val point = ItemStackData(Material.EXPERIENCE_BOTTLE, Function.decoText("アトリビュート"), lore).view()
        attributeMenuCache!!.setItem(8, point)
    }

    fun attributeMenuClick(
        view: InventoryView,
        clickType: ClickType,
        currentItem: ItemStack,
    ) {
        if (!Function.equalInv(view, Data.AttributeMenuDisplay)) return

        val attr = playerData.Attribute
        for (attrType in AttributeType.entries) {
            if (currentItem.type != attrType.icon) continue

            val x = if (clickType.isShiftClick) 10 else 1
            if (clickType.isLeftClick) {
                attr.addAttribute(attrType, x)
            } else if (clickType.isRightClick) {
                if (playerData.Map.Safe) {
                    attr.revAttribute(attrType, x)
                } else {
                    Function.sendMessage(player, "§eセーフゾーン§aでのみ使用可能です", SoundList.NOPE)
                }
            }
        }
        if (currentItem.type == Material.EXPERIENCE_BOTTLE) {
            if (playerData.Map.Safe) {
                attr.resetAttribute()
            } else {
                Function.sendMessage(player, "§eセーフゾーン§aでのみ使用可能です", SoundList.NOPE)
            }
        }
        attributeMenuLoad()
        CustomSound.playSound(player, SoundList.CLICK)
    }
}
